### PR TITLE
[WIP] ENH: Add draft coordinate image API as nilearn.coordimage

### PR DESCRIPTION
**Pull request by effigies**
I've done some very cursory edits to adjust imports, but this is more-or-less the significant portions of https://github.com/nipy/nibabel/pull/1090.

I don't have full test coverage, and I'm quite sure that the parts that depend on nibabel test directories will need to be changed, but let's see how broken it is...

Touches on:
* #2171
* #2388
* #2423
* #2681 
* and probably others...

Changes proposed in this pull request:

- Adds `coordimage.pointset` to encode descriptions of geometry
  - Triangular meshes and ND grids are implemented
  - `TriMeshFamily` is intended as a drop-in replacement for `TriangularMesh` that adds a "name" parameter to most functions to make it easy to shift to alternative coordinates in the same mesh.
- Adds `coordimage.coordimage`
  - `CoordinateImage`'s `(data, coordaxis, header)` model is intended to mirror `SpatialImage`'s `(data, affine, header)`
  - A coordinate axis is a series of parcels
  - A parcel is a (subsampled) geometry
- Adds `CaretSpecFile` to encode Connectome Workbench `wbspec` files, which associate CIFTI-2 files with geometry files

### CODE REVIEW

**Comment by jeromedockes** [comment]
Thanks so much for getting started on this! 
- as this is provisional I don't think having full test coverage should be a blocker at this point -- besides, the rest of nilearn doesn't have it either
- I wonder if for now we should name the module `_coordimage`? or make it clear in some other way that this is WIP and subject to change?
- I think we may be missing an `__init__.py` in `coordimage`
- ATM data files are listed [ in `setup.py`](https://github.com/nilearn/nilearn/blob/main/setup.py#L69)

**Comment by ymzayek** [comment]
Hi @effigies, when you get the chance can you clarify the next steps with this PR?
Also, the logic for loading the different surface file formats is not clear to me. Can you provide some examples?

**Comment by effigies** [comment]
Hi, thanks. Sorry I will try to get this going a bit better in the next week.

**Review by htwangtw — COMMENTED** [review]
Just questions related to `CaretSpecFile`. I am not familiar with it or xml files in general.

File: nilearn/coordimage/caretspec.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        # Collecting char buffer fragments
+        self._char_blocks = []
+
+    def StartElementHandler(self, name, attrs):
+        self.flush_chardata()
+        if name == 'CaretSpecFile':
+            self.caret_spec = CaretSpecFile(version=attrs['Version'])
+        elif name == 'MetaData':
+            self.caret_spec.metadata = CaretMetaData()
+        elif name == 'MD':
```
[the remark this answers is not in the document]
Comment (effigies) [reply nilearn/coordimage/caretspec.py:161]: It's amazing that people prefer JSON over XML...

**Comment by effigies** [comment]
I did not get this going a few weeks ago... I'm very sorry. I haven't forgotten about this.

**Comment by ymzayek** [comment]
Hi @effigies would you mind if I or one of the nilearn developers try to take this PR over? We will discuss it tomorrow during our core developers meeting at 4pm UTC. You are of course more than welcome to join if you find the time!

**Comment by effigies** [comment]
Hi, yeah that's totally fine. I'm very sorry for how long it's taken me to get back to this, and if you have the bandwidth to work on it, please go ahead.

**Comment by Remi-Gau** [comment]
Converting to draft PR for now

**Comment by bthirion** [comment]
Isn't it superseded by  #3856  ?

**Comment by effigies** [comment]
Yeah, let's not have this here. If there are useful bits to pull out of it, let's scope it out in an issue and make more targeted PRs.

**Pull request closed without merging**